src/Microsoft.PowerShell.ConsoleGuiTools/ConsoleGui.cs

```diff
@@ -56,9 +56,46 @@ public void Start(ApplicationData applicationData)
             top.Add(menu);
 
             var gridHeaders = applicationData.DataTable.DataColumns.Select((c) => c.Label).ToList();
-            // We add one as the offset here to line it up with the data.
-            win.Add(new Label(GetPaddedString(gridHeaders, top.Frame.Width - 3, offset: 1)));
-            
+            var columnWidths = new int[gridHeaders.Count];
+            int index = 0;
+            foreach (var header in gridHeaders)
+            {
+                columnWidths[index] = header.Length;
+                index++;
+            }
+
+
+            // calculate the width of each column based on longest string in each column for each row
+            foreach (var row in applicationData.DataTable.Data)
+            {
+                index = 0;
+                foreach (var col in row.Values)
+                {
+                    var len = col.Value.DisplayValue.Length;
+                    if (len > columnWidths[index])
+                    {
+                        columnWidths[index] = len;
+                    }
+                    
+                    index++;
+                }
+            }
+
+            // If we have PassThru, then we want to make them selectable. If we make them selectable,
+            // they have a 8 character addition of a checkbox ("     [ ]") that we have to factor in.
+            int offset = applicationData.PassThru ? 8 : 4;
+
+            // if the total width is wider than the usable width, remove 1 from widest column until it fits
+            // the gui loses 3 chars on the left and 2 chars on the right
+            int usableWidth = top.Frame.Width - 3 - columnWidths.Length - offset - 2;
+            while (columnWidths.Sum() >= usableWidth)
```

Comment:
Do you need to calculate the sum each time? You're only subtracting 1 each time. This could just be a for loop.

Reply:
Good point, since it's just decrementing by one, I could do the same for the sum.